File: my-project/src/algoritma/AutoKeyVigenereCipher.js
function VigenereCipherEncrypt (plaintext, key){
    // sanitize plaintext delete spaces
    plaintext = plaintext.replace(/\s/g, "");
    // uppercase plaintext and key
    plaintext = plaintext.toUpperCase();
    key = key.toUpperCase();
    let ciphertext = "";
    let keyIndex = 0;
    for (var i = 0; i < plaintext.length; i++){
        var charCode = plaintext.charCodeAt(i);
        if (charCode >= 65 && charCode <= 90){
            ciphertext += String.fromCharCode((charCode - 65 + key.charCodeAt(keyIndex) - 65) % 26 + 65);
            keyIndex = (keyIndex + 1) % key.length;
        }
        else if (charCode >= 97 && charCode <= 122){
            ciphertext += String.fromCharCode((charCode - 97 + key.charCodeAt(keyIndex) - 97) % 26 + 97);
            keyIndex = (keyIndex + 1) % key.length;
        }
        else {
            ciphertext += plaintext[i];
        }
    }
    // divide ciphertext into 5 characters
    ciphertext = ciphertext.match(/.{1,5}/g).join(" ");

    return ciphertext;
}
function autoKeyVigenereCipherEncrypt (plaintext, key) {
    // sanitize plaintext delete spaces
    plaintext = plaintext.replace(/\s/g, "");
    // santize key delete spaces
    key = key.replace(/\s/g, "");
    let ciphertext = "";
    if (plaintext.length > key.length) {
        key = key + plaintext.substring(0, plaintext.length - key.length);
    }
    ciphertext = VigenereCipherEncrypt(plaintext, key);
    return ciphertext;
}

function VigenereCipherDecrypt (ciphertext, key){
    // sanitize ciphertext delete spaces
    ciphertext = ciphertext.replace(/\s/g, "");
    // uppercase ciphertext and key
    ciphertext = ciphertext.toUpperCase();
    key = key.toUpperCase();
    let plaintext = "";
    let keyIndex = 0;
    for (var i = 0; i < ciphertext.length; i++){
        var charCode = ciphertext.charCodeAt(i);
        if (charCode >= 65 && charCode <= 90){
            plaintext += String.fromCharCode((charCode - 65 - key.charCodeAt(keyIndex) + 65 + 26) % 26 + 65);
            console.log("plaintext: " + plaintext);
            keyIndex = (keyIndex + 1) % key.length;
        }
        else {
            plaintext += ciphertext[i];
        }
        if (key.length < ciphertext.length){
            key += plaintext[i];
        }
    }
    // divide plaintext into 5 characters
    plaintext = plaintext.match(/.{1,5}/g).join(" ");

    return plaintext;
}
function autoKeyVigenereCipherDecrypt (ciphertext, key) {
    // sanitize ciphertext delete spaces
    ciphertext = ciphertext.replace(/\s/g, "");
    // santize key delete spaces
    let plaintext = "";
    console.log ("key: " + key);
    plaintext = VigenereCipherDecrypt(ciphertext, key);
    return plaintext;
}

function main(){
    let plaintext;
    plaintext = "negara penghasil minyak";
    let key;
    key = "INDO";

    let ciphertext;
    ciphertext = autoKeyVigenereCipherEncrypt(plaintext, key);
    console.log("ciphertext: " + ciphertext);

    let plaintextBaru = autoKeyVigenereCipherDecrypt(ciphertext, key);
    console.log("plaintext: "+ plaintextBaru);
}

main();

File: api-cipher/algorithm/autoKeyVigenereCipher.py
def autoKeyVigenereCipherEncrypt(plaintext, key, group):
    plaintext = plaintext.replace(" ", "")
    key = key.replace(" ", "")
    plaintext = plaintext.upper()
    key = key.upper()
    ciphertext = ""
    if len(key) < len(plaintext):
        key += plaintext[:len(plaintext) - len(key)]
    for i in range(len(plaintext)):
        ciphertext += chr((ord(plaintext[i]) + ord(key[i])) % 26 + 65)
    if group == True:
        plaintextGroup = ""
        for i in range(len(plaintext)):
            plaintextGroup += plaintext[i]
            if (i+1) % 5 == 0:
                plaintextGroup += " "
        return plaintextGroup
    return ciphertext

def autoKeyVigenereCipherDecrypt(ciphertext, key):
    ciphertext = ciphertext.replace(" ", "")
    key = key.replace(" ", "")
    ciphertext = ciphertext.upper()
    key = key.upper()
    plaintext = ""
    keyIndex = 0
    for i in range(len(ciphertext)):
        plaintext += chr((ord(ciphertext[i]) - ord(key[i])) % 26 + 65)
        key += plaintext[i]

    return plaintext


if __name__ == "__main__":
    plaintext = "negarapenghasilminyak"
    key = "INDO"
    ciphertext = autoKeyVigenereCipherEncrypt(plaintext, key)
    plaintext = autoKeyVigenereCipherDecrypt(ciphertext, key)
    print("ciphertext", ciphertext)
    print("plaintext", plaintext)
